sous-chef/sous_chef/grocery_list/generate_grocery_list.py

```diff
@@ -202,18 +201,101 @@ def _add_menu_recipe_to_queue(self, menu_recipe_list: list[MenuRecipe]):
             self.queue_menu_recipe = []
         self.queue_menu_recipe.extend(menu_recipe_list)
 
+    def _add_preparation_task_to_queue(
+        self,
+        task: str,
+        weekday: str,
+        hour: int,
+        minute: int,
+        from_recipe: str,
+        from_day: str,
+    ):
+        preparation_task = pd.DataFrame(
+            {
+                "task": [task],
+                "weekday": [weekday],
+                "hour": [hour],
+                "minute": [minute],
+                "from_recipe": [from_recipe],
+                "from_day": from_day,
+            }
+        )
+        if self.queue_preparation is None:
+            self.queue_preparation = preparation_task
+        self.queue_preparation = pd.concat(
+            [self.queue_preparation, preparation_task]
+        )
+
     def _add_referenced_recipe_to_queue(
         self, menu_recipe: MenuRecipe, recipe_list: List[Recipe]
     ):
+        def _check_yes_no(text: str) -> str:
+            response = None
+            while response not in ["Y", "N"]:
+                response = input(f"\n{text} [Y/N] ").upper()
+            return response
+
+        def _get_schedule_day_hour_minute():
+            day = None
+            while day not in Weekday.name_list("capitalize"):
+                day = input("\nWeekday: ").capitalize()
+
+            meal_time = None
+            meal_times = MealTime.name_list("lower")
+            while meal_time not in meal_times:
+                meal_time = input(f"\nMealtime {meal_times}: ").lower()
+            meal_time = MealTime[meal_time].value
+            return day, meal_time["hour"], meal_time["minute"]
+
+        def _give_referenced_recipe_details():
+            total_factor = menu_recipe.eat_factor + menu_recipe.freeze_factor
+            print("\n#### REFERENCED RECIPE CHECK ####")
+            print(f"- from recipe: {menu_recipe.recipe.title}")
+            print(f"-- total_factor: {total_factor}")
+            print(f"-- from day: {menu_recipe.from_day}")
+            print(f"- referenced recipe: {recipe.title}")
+            print(f"-- amount needed: {recipe.amount}")
+            print(f"-- total time: {recipe.total_cook_time}")
+
         for recipe in recipe_list:
-            menu_recipe = MenuRecipe(
-                from_recipe=f"{recipe.title}_{menu_recipe.recipe.title}",
+            from_recipe = f"{recipe.title}_{menu_recipe.recipe.title}"
+            menu_sub_recipe = MenuRecipe(
+                from_recipe=from_recipe,
                 from_day=menu_recipe.from_day,
                 eat_factor=menu_recipe.eat_factor * recipe.factor,
                 freeze_factor=menu_recipe.freeze_factor * recipe.factor,
                 recipe=recipe,
             )
-            self._add_menu_recipe_to_queue([menu_recipe])
+
+            # TODO make more robust to other methods
+            if self.config.run_mode.with_todoist:
```

Comment:
Add tests to cover all this new logic

Reply:
could do that, not sure how critical. what do you think?